service/frontend/workflowHandler.go

```diff
@@ -1622,6 +1622,36 @@ func (wh *WorkflowHandler) RespondActivityTaskFailed(
 	sizeLimitError := wh.config.BlobSizeLimitError(namespaceEntry.Name().String())
 	sizeLimitWarn := wh.config.BlobSizeLimitWarn(namespaceEntry.Name().String())
 
+	if request.GetLastHeartbeatDetails() != nil {
+		if err := common.CheckEventBlobSizeLimit(
+			request.GetLastHeartbeatDetails().Size(),
+			sizeLimitWarn,
+			sizeLimitError,
+			namespaceID.String(),
+			taskToken.GetWorkflowId(),
+			taskToken.GetRunId(),
+			wh.metricsScope(ctx).Tagged(metrics.CommandTypeTag(enumspb.COMMAND_TYPE_UNSPECIFIED.String())),
+			wh.throttledLogger,
+			tag.BlobSizeViolationOperation("RespondActivityTaskFailed"),
+		); err != nil {
+			// heartbeat details exceed size limit, we would fail the activity immediately with explicit error reason
+			failRequest := &workflowservice.RespondActivityTaskFailedRequest{
+				TaskToken: request.TaskToken,
+				Failure:   failure.NewServerFailure(common.FailureReasonHeartbeatExceedsLimit, true),
+				Identity:  request.Identity,
+			}
```

Comment:
Instead of drop the actual failure reason and fail activity with the size exceed limit, i would rather drop the over sized heartbeat, and record the actual failure.

Reply:
For the case of the activity retry on failure and using heartbeat to be idempotent that would cause a potential "silent" error.

This models the same behavior as if the activity sends the heartbeat in a separate api call.

The trade off here to me seemed correct, but I do want to understand this choice better.

Reply:
idempotence reference was about how we document SDK consumers can use the heartbeat details to achieve idempotence.

https://docs.temporal.io/docs/go/how-to-heartbeat-an-activity-in-go

I believe that a more detailed error/response is the path I am going to need to go down to provide the SDK with the necessary details for the consumer.

Reply:
> I believe that a more detailed error/response is the path I am going to need to go down to provide the SDK with the necessary details for the consumer.

+1 to this

Reply:
Updated, now the heartbeat would be dropped on size limit violation.
Response now includes failures if any encountered during input validation.